src/native/datastream/data_stream.h

```diff
@@ -0,0 +1,174 @@
+#ifdef _MSC_VER
+#define DATA_STREAM_EXPORT __declspec(dllexport)
+#else
+#define DATA_STREAM_EXPORT __attribute__ ((visibility ("default")))
+#endif // _MSC_VER
+#else
+#define DATA_STREAM_EXPORT
+#endif // BUILD_SHARED_LIBRARY
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif // __cplusplus
+
+// Forward declarations
+typedef struct data_stream__ data_stream_t;
+
+typedef struct data_stream_context__
+{
+    uint32_t magic; // Value used to confirm the pointer and check endianness of the target machine.
+    uint16_t size; // The current size of this data structure.
+    uint16_t version; // The version of the data stream binary format.
+
+    uint32_t reserved; // Reserved.
+    uint32_t streams_count; // Count of statically allocated streams.
+    data_stream_t* streams; // Statically allocated streams.
+} data_stream_context_t;
+
+// Validation results for a data stream.
+enum
+{
+    dsv_invalid,
+    dsv_little_endian,
+    dsv_big_endian,
+};
+typedef int32_t ds_validate_t;
+
+// Validate the magic number is valid
+ds_validate_t dnds_validate(uint32_t magic);
+
+// Boolean return type for data stream functions.
+typedef unsigned char ds_bool;
+
+// Determine if the data stream is big or little endian.
+DATA_STREAM_EXPORT ds_bool dnds_is_big_endian(data_stream_context_t*);
+
+//
+// Target APIs
+//
+
+// Initialize the data stream.
+// Statically allocates streams with a default block size.
+DATA_STREAM_EXPORT ds_bool dnds_init(
+    data_stream_context_t*,
+    uint32_t stream_count,
+    size_t const* stream_byte_lengths);
+
+// Deallocate all memory associated with the data stream.
+DATA_STREAM_EXPORT void dnds_destroy(data_stream_context_t*);
+
+// Define a maximum value for the type datatype.
+#define MAX_TYPE_SIZE UINT16_MAX
+
+typedef struct type_details__
+{
+    uint16_t type;
+    uint16_t version;
+    uint16_t reserved; // Must be zero
+    uint16_t name_len; // Includes null
+    char const* name;
+} type_details_t;
+static_assert((offsetof(type_details_t, name) % sizeof(void*)) == 0, "Pointer field should be pointer aligned");
+
+typedef struct field_offset__
+{
+    uint16_t offset;
+    uint16_t type;
+} field_offset_t;
+static_assert(sizeof(field_offset_t) == sizeof(uint32_t), "Field offset structures should be 4 bytes");
+
+// Define a type in the data stream
+DATA_STREAM_EXPORT ds_bool dnds_define_type(
+    data_stream_context_t*,
+    type_details_t const* details,
+    size_t total_size,
+    size_t offsets_length,
+    field_offset_t const* offsets);
+
+// Get an available stream to write to
+DATA_STREAM_EXPORT data_stream_t* dnds_get_stream(
+    data_stream_context_t*,
+    size_t id);
+
+// Record an instance in the stream
+DATA_STREAM_EXPORT ds_bool dnds_record_instance(
+    data_stream_t*,
+    uint16_t type,
+    void* inst);
+
+// Record a data blob in the stream
+DATA_STREAM_EXPORT ds_bool dnds_record_blob(
+    data_stream_t*,
+    uint16_t type,
+    uint16_t size,
+    void* inst);
+
+//
+// Reader APIs
+//
+
```

Comment:
I think the reader APIs have some type confusion between the host and the debuggee target.

* `data_stream_context_t:streams` is a `data_stream_t*` which is a host pointer-sized value.  That seems wrong. It should be a target pointer (so in particular it might be 32-bit BE).

* `memory_reader_t:read_ptr` is a function like `ds_bool read_remote(memory_reader_t *self, intptr_t remote_addr, size_t* count, void **dest)`.  Here `remote_addr` is not `intptr_t` it's an opaque target pointer

If we only ever plan to run the reader on a 64-bit arch, this is probably good enough (except we should be careful when we read `data_stream_context_t` from the remote end that we don't read more than `count` bytes)

I think we should use `typedef uint64_t ds_foreignptr_t` and store remote addresses as opaque 64-bit blobs.   

We should probably use a different `data_stream_remote_context_t` struct for all the reader APIs (with a `ds_foreignptr_t` for the `streams` field) and a the `remote_addr` param of `memory_reader_t:read_ptr` should also be a `ds_foreignptr_t`.

Reply:
Actually there's more 64-bit host / 32-bit target issues:  `read_in_block_data` has ` size_t remote_block_read = sizeof(data_block_t);` which is using the host architecture's sizeof for the data block (which has fields of host-architecture pointer types) to read in target data.